### DISCUSSION TITLE
`diffedit3`, a prototype 3-pane diffediting GUI/webapp

### DISCUSSION THREAD

**Discussion by ilyagr** [jj-vcs/jj#3094] [Show and tell]
I made ["diffedit3", a prototype GUI diff editor for `jj`](https://github.com/ilyagr/diffedit3) that I'm hoping to bundle with `jj` eventually (we might then rename it). It's a lowest-common-denominator kind of thing: a web app that works from a local server `jj` would start, and looks a lot like a much less polished version of `meld`. The idea is that it would work over SSH and wouldn't need the user to install anything.

Currently, there is also a version of it that's a full-fledged GUI application, based on Tauri.

It competes with `scm-diff-editor` a little, but I think there's a need for both a TUI and a GUI. I feel like 3-pane diffediting adds something essential to the `jj` experience that I want to be accessible, and it seems easiest to make this work in the JS world. In the case of `diffedit3`, it's a thin wrapper around CodeMirror5's merge plugin (which comes with some issues, see the Readme).

Take a look, let me know what you think! There should be binaries for all platforms on the releases page.

![image](https://github.com/martinvonz/jj/assets/4123047/f918027a-dc06-4ec6-954d-eec348037365)

**Comment by joyously** [comment]
It's Python, but there is [Diffuse](https://github.com/MightyCreak/diffuse) that does multi pane diff editing.
I recently came across [Textual](https://www.textualize.io/) for TUI framework. It's Python also, but seems pretty nice.

**Reply by ilyagr** [reply]
I've recently learned about Diffuse myself. It does what I want, but was extremely painful to install on Mac IIRC, and seems a bit less polished than Meld.

Textual seems nice, but I think making a diff pane UI with it would still be a lot of work. It might not be for me, I find the idea of using CSS to set up a TUI weird. I wouldn't enjoy learning the differences between CSS and Textual CSS. Finally, for use with `jj`, rust seems better.

As a bit of an aside, for a while I was thinking of trying to make this kind of diff editing work with [`micro`](https://github.com/zyedidia/micro) as a "TUI framework". It'd be like vimdiff, but without requiring people to get familiar with Vim. However, this began to seem like a ton of work after I thought about it and implemented [a small first step](https://github.com/zyedidia/micro/pull/2759). Also, unfortunately, that's a single-author project where the author seems to be focusing on other things these days. Also, that's Go, not Rust (though I think we *could* embed a Go program in jj if we really, really wanted to; it can be statically linked and should work on all platforms that jj would work on without too much trouble).

**Comment by ilyagr** [comment]
Updates: There's a new version of `diffedit3` at https://github.com/ilyagr/diffedit3. Also, I have a first draft of a PR merging it into `jj` at https://github.com/martinvonz/jj/pull/3292. Future discussion is likely to happen there.
